### PR TITLE
tools/litex_sim: add support for other sdram types

### LINKED ISSUE #355: litex_sim: add LPDDR/DDR/DDR2/DDR3/DDR4 support
Requires https://github.com/enjoy-digital/litex/issues/352 and https://github.com/enjoy-digital/litex/issues/353.

Each SDRAM module has a `memtype` attribute that can be used to identify the type of SDRAM module.

For each type of memory, `phy_settings` will be created based on settings already used on hardware:
- DDR/LPDDR: Pipistrello settings.
- DDR2: Nexys4DDR settings.
- DDR3: Arty settings.

To get the inital `phy_settings` values, some traces can be added to the LiteDRAM PHY:
- https://github.com/enjoy-digital/litedram/blob/master/litedram/phy/s6ddrphy.py#L63
- https://github.com/enjoy-digital/litedram/blob/master/litedram/phy/s7ddrphy.py#L71

Once simulation is working with fixed values for each configuration, computation done inside the PHY to generate the `phy_settings` should be integrated in the simulation.

While doing this work, doing first point of https://github.com/enjoy-digital/litedram/issues/60 would also be useful to avoid duplicating code for the  `phy_settings` computation.

This change will allow simulating the various SDRAM configurations we are already using on hardware and will ease benchmarking memory accesses with LiteX/LiteDRAM.

### PR DESCRIPTION
**Pull request by piotr-binkowski**
This PR will allow litex_sim to run simulations using other types of sdram memory (DDR, LPDDR, DDR2, DDR3) instead of being limited to SDR only

Closes #355

### CODE REVIEW

File: litex/tools/litex_sim.py [hunk 1]
```diff
@@ -62,6 +62,72 @@ class Platform(SimPlatform):
...
+    "nphases": 1,
+    "rdphase": 0,
+    "wrphase": 0,
+    "rdcmdphase": 0,
+    "wrcmdphase": 0,
+    "cl": 2,
+    "read_latency": 4,
+    "write_latency": 0
+}
+
```
Comment (enjoy-digital) [inline litex/tools/litex_sim.py]: This should also be integrated in get_sdram_phy_settings.

File: litex/tools/litex_sim.py [hunk 2]
```diff
@@ -62,6 +62,72 @@ class Platform(SimPlatform):
...
+sdram_module_rates = {
+    "SDR": "1:1",
+    "DDR": "1:2",
+    "LPDDR": "1:2",
+    "DDR2": "1:2",
+    "DDR3": "1:4",
+}
+
+def get_sdram_phy_settings(memtype, data_width, clk_freq):
+    nphases = 4 if memtype == "DDR3" else 2
```
Comment (enjoy-digital) [inline litex/tools/litex_sim.py]: `rates`/`nphases` values always match so you can probably also reuse `sdram_module_rates` dict.

File: litex/tools/litex_sim.py [hunk 3]
```diff
@@ -62,6 +62,72 @@ class Platform(SimPlatform):
...
+
+    # Settings taken from s7ddrphy
+    if memtype in ["DDR2", "DDR3"]:
+        tck                 = 2/(2*nphases*clk_freq)
```
Comment (enjoy-digital) [inline litex/tools/litex_sim.py]: We should probably organize things in the order of the complexity: SDR, then DDR/LPDDR, then DDR2/DDR3.

File: litex/tools/litex_sim.py [hunk 4]
```diff
@@ -62,6 +62,72 @@ class Platform(SimPlatform):
...
+        "nphases": nphases,
+        "rdphase": rdphase,
+        "wrphase": wrphase,
+        "rdcmdphase": rdcmdphase,
+        "wrcmdphase": wrcmdphase,
+        "cl": cl,
+        "cwl": cwl,
+        "read_latency": read_latency,
+        "write_latency": write_latency,
+    } if memtype != "SDR" else sdr_phy_settings
```
Comment (enjoy-digital) [inline litex/tools/litex_sim.py]: The if can be removed when `sdr_phy_settings` will be generated in `get_sdram_phy_settings`.

**Comment by piotr-binkowski** [comment]
I've got memtest to pass only for DDR/LPDDR at this point, calculated settings for DDR2 and DDR3 are matching what I got from running `Arty` and `Nexys4DDR` targets but memtest fails for them.
ie. when I run: `./tools/litex_sim.py --with-sdram --sdram-module=MT47H64M16 --sdram-data-width=16` which matches Nexys4DDR I get:
```
 BIOS built on Jan 24 2020 11:55:54
 BIOS CRC passed (ff70f62a)

 Migen git sha1: d11565a
 LiteX git sha1: 64e9fada

--=============== SoC ==================--
CPU:       VexRiscv @ 1MHz
ROM:       32KB
SRAM:      4KB
L2:        8KB
MAIN-RAM:  131072KB

--========== Initialization ============--
Initializing SDRAM...
SDRAM now under hardware control
Memtest bus failed: 128/256 errors
Memtest data failed: 2048/2048 errors
Memtest addr failed: 2048/2048 errors
Memory initialization failed
```

**Comment by enjoy-digital** [comment]
Ok thanks. The issue could also be related to the DFI model. If you want to have more information, you can enable debug by adding the following constants to the SoC:
````
            self.add_constant("MEMTEST_BUG_DEBUG", None)
            self.add_constant("MEMTEST_ADDR_DEBUG", None)
            self.add_constant("MEMTEST_DATA_DEBUG", None)
````

File: litex/tools/litex_sim.py (hunk 1 above)
Comment (piotr-binkowski) [reply litex/tools/litex_sim.py]: This is now part of `get_sdram_phy_settings`

File: litex/tools/litex_sim.py (hunk 2 above)
Comment (piotr-binkowski) [reply litex/tools/litex_sim.py]: Good idea, I'm now using single dict for `rates/nphases`

File: litex/tools/litex_sim.py (hunk 3 above)
Comment (piotr-binkowski) [reply litex/tools/litex_sim.py]: This is in correct order now

File: litex/tools/litex_sim.py (hunk 4 above)
Comment (piotr-binkowski) [reply litex/tools/litex_sim.py]: Removed as it is no longer needed

**Comment by piotr-binkowski** [comment]
I've enabled debug and it looks like `bus` test fails on half of the addresses
```
SDRAM now under hardware control
[bus: 0x2]: 0x00000000 vs 0xaaaaaaaa
[bus: 0x3]: 0x00000000 vs 0xaaaaaaaa
[bus: 0x6]: 0x00000000 vs 0xaaaaaaaa
[bus: 0x7]: 0x00000000 vs 0xaaaaaaaa
[bus: 0xa]: 0x00000000 vs 0xaaaaaaaa
[bus: 0xb]: 0x00000000 vs 0xaaaaaaaa
[bus: 0xe]: 0x00000000 vs 0xaaaaaaaa
[bus: 0xf]: 0x00000000 vs 0xaaaaaaaa
...
```
And data in `data` and `addr` tests seems to be missing in half of the cases and visible on a different address in the remaining cases
```
Memtest bus failed: 128/256 errors
[data 0x0]: 0xd1ccf6e9 vs 0x3c6ef35f
[data 0x1]: 0xaaf95334 vs 0x47502932
[data 0x2]: 0x00000000 vs 0xd1ccf6e9
[data 0x3]: 0x00000000 vs 0xaaf95334
[data 0x4]: 0x57fe6c2d vs 0x6252e503
[data 0x5]: 0xa3d95fa8 vs 0x9f2ec686
[data 0x6]: 0x00000000 vs 0x57fe6c2d
[data 0x7]: 0x00000000 vs 0xa3d95fa8
[data 0x8]: 0xcbf633b1 vs 0x81fdbee7
[data 0x9]: 0xbcd1195c vs 0x94f0af1a
[data 0xa]: 0x00000000 vs 0xcbf633b1
[data 0xb]: 0x00000000 vs 0xbcd1195c
[data 0xc]: 0x01ba5175 vs 0x9d23e50b
[data 0xd]: 0x83c6b450 vs 0xe296f6ee
[data 0xe]: 0x00000000 vs 0x01ba5175
[data 0xf]: 0x00000000 vs 0x83c6b450
```
```
Memtest data failed: 2048/2048 errors
[addr 0x0]: 0x00000002 vs 0x00000000
[addr 0x1]: 0x00000000 vs 0x00000001
[addr 0x2]: 0x00000000 vs 0x00000002
[addr 0x3]: 0x00000005 vs 0x00000003
[addr 0x4]: 0x00000006 vs 0x00000004
[addr 0x5]: 0x00000000 vs 0x00000005
[addr 0x6]: 0x00000000 vs 0x00000006
[addr 0x7]: 0x00000009 vs 0x00000007
[addr 0x8]: 0x0000000a vs 0x00000008
[addr 0x9]: 0x00000000 vs 0x00000009
[addr 0xa]: 0x00000000 vs 0x0000000a
[addr 0xb]: 0x0000000d vs 0x0000000b
[addr 0xc]: 0x0000000e vs 0x0000000c
[addr 0xd]: 0x00000000 vs 0x0000000d
[addr 0xe]: 0x00000000 vs 0x0000000e
[addr 0xf]: 0x00000011 vs 0x0000000f
```

**Comment by enjoy-digital** [comment]
Thanks for the results, it can be a missmatch on one of the parameters or an issue with the model. We will probably need to use `--trace` to look at the .vcd to understand. (i could look at that later unless you want to).

**Comment by piotr-binkowski** [comment]
I did some more tests with `--sdram-data-width=8` and it looks like this influences the results:
```
Memtest bus failed: 64/256 errors
[data 0x0]: 0x47502932 vs 0x3c6ef35f
[data 0x1]: 0xd1ccf6e9 vs 0x47502932
[data 0x2]: 0xaaf95334 vs 0xd1ccf6e9
[data 0x3]: 0x00000000 vs 0xaaf95334
[data 0x4]: 0x9f2ec686 vs 0x6252e503
[data 0x5]: 0x57fe6c2d vs 0x9f2ec686
[data 0x6]: 0xa3d95fa8 vs 0x57fe6c2d
[data 0x7]: 0x00000000 vs 0xa3d95fa8
[data 0x8]: 0x94f0af1a vs 0x81fdbee7
[data 0x9]: 0xcbf633b1 vs 0x94f0af1a
[data 0xa]: 0xbcd1195c vs 0xcbf633b1
[data 0xb]: 0x00000000 vs 0xbcd1195c
[data 0xc]: 0xe296f6ee vs 0x9d23e50b
[data 0xd]: 0x01ba5175 vs 0xe296f6ee
[data 0xe]: 0x83c6b450 vs 0x01ba5175
[data 0xf]: 0x00000000 vs 0x83c6b450
```
Now data looks to be shifted by one word and only 1 out of 4 words is missing.
It would be great if you could take a look at the `.vcd` outputs.

**Comment by enjoy-digital** [comment]
Yes i will look at that later. That's fine for the integration of the different SDRAM module, so I'll merge your PR but will keep #355 open for now.

**Pull request merged**